test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java

```diff
@@ -0,0 +1,224 @@
+                    entry.setComment(LAST_CEN_COMMENT);
+                }
+                // Add the entry
+                zip.putNextEntry(entry);
+
+
+            }
+            // Finish writing the last entry
+            zip.closeEntry();
+
+            // Before the CEN headers are written, set the extra data on each entry
+            byte[] extra = makeLargeExtraField();
+            for (ZipEntry entry : entries) {
+                entry.setExtra(extra);
+            }
+        }
+    }
+
+    /**
+     * Validates that a ZipException is thrown with the expected message when
+     * the ZipFile is initialized with a ZIP whose CEN exeeds {@link #MAX_CEN_SIZE}
+     */
+    @Test
+    public void centralDirectoryTooLargeToFitInByteArray() {
+        ZipException ex = assertThrows(ZipException.class, () -> new ZipFile(hugeZipFile));
+        assertEquals(CEN_TOO_LARGE_MESSAGE, ex.getMessage());
+    }
+
+    /**
+     * We can reduce the number of written CEN headers by making each CEN header maximally large.
+     * We do this by adding the extra field produced by this method to each CEN header.
+     * <p>
+     * The structure of an extra field is as follows:
+     * <p>
+     * Header ID  (Two bytes, describes the type of the field, also called 'tag')
+     * Data Size  (Two byte short)
+     * Data Block (Contents depend on field type)
+     */
+    private byte[] makeLargeExtraField() {
+        // Make a maximally sized extra field
+        byte[] extra = new byte[MAX_EXTRA_FIELD_SIZE];
+        // Little-endian ByteBuffer for updating the header fields
+        ByteBuffer buffer = ByteBuffer.wrap(extra).order(ByteOrder.LITTLE_ENDIAN);
+
+        // We use the 'unknown' tag, specified in APPNOTE.TXT, 4.6.1 Third party mappings'
+        buffer.putShort(UNKNOWN_ZIP_TAG);
+
+        // Size of the actual (empty) data
+        buffer.putShort(MAX_DATA_SIZE);
+        return extra;
+    }
+
+    /**
+     * By writing sparse 'holes' until the last CEN is detected, we can save disk space
+     * used by this test from ~2GB to ~4K
+     */
+    private static class SparseOutputStream extends FilterOutputStream {
+        private final FileChannel channel;
+        private boolean sparse = true; // True until the last CEN is written
+        private long position = 0;
+
+        public SparseOutputStream(FileOutputStream fos) {
+            super(fos);
+            this.channel = fos.getChannel();
+        }
+
+        @Override
+        public void write(byte[] b, int off, int len) throws IOException {
+            position += len;
+            if (sparse) {
+                // Until finding the last CEN, we don't actually write anything,
+                // but instead simply advance the position, creating a sparse file
+                channel.position(position);
+                // Check for last CEN record
+                if (Arrays.equals(LAST_CEN_COMMENT_BYTES, 0, LAST_CEN_COMMENT_BYTES.length, b, off, len)) {
```

Comment:
The way the instance of a `SparseOutputStream` is used in this test, it gets passed to the constructor of `ZipOutputStream`. That then means that there is no buffering involved when bytes are written out to the output stream internally by `ZipOutputStream`. The implementation of `ZipOutputStream` writes out the `ZipEntry`'s comment (if any) in one single `write(byte[])` call on the outputstream, so it's guaranteed that the `byte[] b` passed in here will actually have the entire comment (from `off` to `len`). So this `Arrays.equals(...)` check is then guaranteed to pass (when that specific entry does get written out). So this check looks good to me.

Reply:
Thanks, this was a nice observation. I added a short comment to the SparseOutputStream noting the instances should be passed directly, without any buffering.